src/extension/portmap/portmap.c

```diff
@@ -0,0 +1,367 @@
+/*
+ * Copyright (C) 2016 Vincent Hage
+ */
+
+#include <stdint.h>         /* intptr_t, */
+#include <stdlib.h>         /* strtoul(3), */
+#include <sys/un.h>		    /* strncpy */
+#include <sys/socket.h>	    /* AF_UNIX, AF_INET */
+#include <arpa/inet.h>      /* inet_ntop */
+
+#include "cli/note.h"
+#include "extension/extension.h"
+#include "tracee/mem.h"				/* read_data */
+#include "syscall/chain.h"			/* register_chained_syscall */
+#include "extension/portmap/portmap.h"
+
+Extension *global_portmap_extension = NULL;
+
+typedef struct Config {
+	PortMap portmap;
+	bool netcoop_mode;
+	bool need_to_check_new_port;
+	uint16_t old_port;
+	word_t sockfd;
+} Config;
+
+/**
+ * Change the port of the socket address, if it maps with an entry.
+ * Return 0 if no relevant entry is found, and 1 if the port has been changed.
+ */
+int change_inet_socket_port(Tracee *tracee, Config *config, struct sockaddr_in *sockaddr, bool bind_mode) {
+	uint16_t port_in, port_out;
+
+	port_in = sockaddr->sin_port;
+	port_out = get_port(&config->portmap, port_in);
+
+	if(port_out == PORTMAP_DEFAULT_VALUE) {
+		if (bind_mode && config->netcoop_mode && !config->need_to_check_new_port) {
```

Comment:
TODO: Maybe add here again the check to prevent ports below 1024 to be affected by the port mapping.
These ports (< 1024) require admin rights anyway, and a port that is already at 0 does not need any help from PRoot.